### PR TITLE
Derive rpc(client,server) for EthereumRpc

### LINKED ISSUE #1599: Derive rpc(client, server) for ethereum-rpc
- Use `jsonrpsee` `rpc` macro in `ethereum-rpc/src/lib.rs` for stricter typing and auto client derivation

### PR DESCRIPTION
**Pull request by jfldde**
# Description

- Derive jsonrpsee rpc macro for `EthereumRpc`
- Remove "local" feature and associated code

## Linked Issues
- Fixes #1599

### CODE REVIEW

File: crates/ethereum-rpc/src/lib.rs [hunk 1]
```diff
@@ -51,503 +53,613 @@ pub struct SyncStatus {
...
+//         other: OtherFields::default(),
+//     };
 
-            let (base_fee, suggested_tip) = ethereum.max_fee_per_gas(&mut working_set);
+//     (call_request, gas_price, max_fee_per_gas)
+// }
 
-            suggested_tip + base_fee
-        };
+#[rpc(server, client)]
```
Comment (eyusufatik) [inline crates/ethereum-rpc/src/lib.rs]: let's not generate client, we'd use alloy for that actually

File: crates/ethereum-rpc/src/lib.rs [hunk 2]
```diff
@@ -51,503 +53,613 @@ pub struct SyncStatus {
...
-            let block_number =
-                match evm.get_block_number_by_block_hash(block_hash, &mut working_set) {
-                    Some(block_number) => block_number,
-                    None => {
-                        return Err(EthApiError::HeaderNotFound(block_hash.into()).into());
-                    }
-                };
+    fn eth_accounts(&self) -> RpcResult<Vec<Address>> {
+        Ok(self.ethereum.eth_signer.signers())
+    }
```
Comment (eyusufatik) [inline crates/ethereum-rpc/src/lib.rs]: remove this function please

File: crates/ethereum-rpc/src/lib.rs [hunk 3]
```diff
@@ -51,503 +53,613 @@ pub struct SyncStatus {
...
+
+    async fn eth_get_transaction_by_hash(
+        &self,
+        hash: B256,
+        mempool_only: Option<bool>,
+    ) -> RpcResult<Option<RpcTransaction<AnyNetwork>>> {
+        match mempool_only {
+            Some(true) => {
+                match SequencerRpcClient::eth_get_transaction_by_hash(
```
Comment (eyusufatik) [inline crates/ethereum-rpc/src/lib.rs]: why not self.ethereum.sequencer_client.as_ref().unwrap().eth_get_transaction_by_hash() ?

File: crates/ethereum-rpc/src/lib.rs [hunk 4]
```diff
@@ -556,227 +668,183 @@ fn register_rpc_methods<C: sov_modules_api::Context, Da: DaService>(
...
+                self.ethereum
+                    .subscription_manager
+                    .as_ref()
+                    .unwrap()
+                    .register_new_heads_subscription(subscription)
+                    .await;
+            }
+            "logs" => {
+                let Some(filter) = filter else {
+                    pending.reject(to_eth_rpc_error("Missing filter")).await;
```
Comment (eyusufatik) [inline crates/ethereum-rpc/src/lib.rs]: you should check what other eth clients return for this scenario. probably the error is standardized

**Review by eyusufatik — APPROVED** [review]

**Review by eyusufatik — CHANGES_REQUESTED** [review]
meant to click on request changes last time 😓

File: crates/ethereum-rpc/src/lib.rs (hunk 3 above)
Comment (jfldde) [reply crates/ethereum-rpc/src/lib.rs]: Was conflicting with the client derive above

File: crates/ethereum-rpc/src/lib.rs (hunk 1 above)
Comment (jfldde) [reply crates/ethereum-rpc/src/lib.rs]: Fair point, I'll remove

File: crates/ethereum-rpc/src/lib.rs (hunk 2 above)
Comment (jfldde) [reply crates/ethereum-rpc/src/lib.rs]: Remove or restore behind local flag ?

Comment (eyusufatik) [reply crates/ethereum-rpc/src/lib.rs]: remove. remove the local flag also. we don't have those features right now. and looks like we won't need to

File: crates/ethereum-rpc/src/lib.rs (hunk 4 above)
Comment (jfldde) [reply crates/ethereum-rpc/src/lib.rs]: Should default https://github.com/paradigmxyz/reth/blob/04af67879c6dac9a9984dedc0e512d16b8eebb1f/crates/rpc/rpc/src/eth/pubsub.rs#L112
Fixed [758e31e](https://github.com/chainwayxyz/citrea/pull/1632/commits/758e31e5112940213b2d95ea0d740043998b919a)

File: bin/citrea/src/eth.rs [hunk 5]
```diff
@@ -40,11 +37,3 @@ pub(crate) fn register_ethereum<Da: DaService>(
         .merge(ethereum_rpc)
         .context("Failed to merge Ethereum RPC modules")
 }
-
-// TODO: #840
```
Comment (jfldde) [inline bin/citrea/src/eth.rs:44]: Couldn't find the associated issue

File: crates/ethereum-rpc/src/lib.rs [hunk 6]
```diff
@@ -770,17 +501,12 @@ where
...
-                    return Ok(());
-                };
-
                 let subscription = pending.accept().await?;
                 self.ethereum
                     .subscription_manager
                     .as_ref()
                     .unwrap()
-                    .register_new_logs_subscription(filter, subscription)
+                    .register_new_logs_subscription(filter.unwrap_or_default(), subscription)
```
Comment (eyusufatik) [inline crates/ethereum-rpc/src/lib.rs:509]: does the rpc really default to a filter??

Comment (eyusufatik) [reply crates/ethereum-rpc/src/lib.rs:509]: saw your comment now

**Review by eyusufatik — APPROVED** [review]

**Comment by eyusufatik** [comment]
requested @yaziciahmet 's review because he wrote all parts regarding the subscriptions.

**Review by yaziciahmet — APPROVED** [review]
Perrrrfect!

**Pull request merged**